src/picosvg/svg_types.py

```diff
@@ -23,6 +23,61 @@
 from typing import Generator, Iterable
 
 
+def _explicit_lines_callback(curr_pos, cmd, args, *_):
+    if cmd == "v":
+        args = (0, args[0])
+    elif cmd == "V":
+        args = (curr_pos.x, args[0])
+    elif cmd == "h":
+        args = (args[0], 0)
+    elif cmd == "H":
+        args = (args[0], curr_pos.y)
+    else:
+        return ((cmd, args),)  # nothing changes
+
+    if cmd.islower():
+        cmd = "l"
+    else:
+        cmd = "L"
+
+    return ((cmd, args),)
+
+
+def _next_pos(curr_pos, cmd, cmd_args):
+    # update current position
+    x_coord_idxs, y_coord_idxs = svg_meta.cmd_coords(cmd)
+    new_x, new_y = curr_pos
+    if cmd.isupper():
+        if x_coord_idxs:
+            new_x = 0
+        if y_coord_idxs:
+            new_y = 0
+
+    if x_coord_idxs:
+        new_x += cmd_args[x_coord_idxs[-1]]
+    if y_coord_idxs:
+        new_y += cmd_args[y_coord_idxs[-1]]
+
+    return Point(new_x, new_y)
+
+
+def _move_endpoint(curr_pos, cmd, cmd_args, new_endpoint):
+    # we need to be able to alter both axes
+    ((cmd, cmd_args),) = _explicit_lines_callback(curr_pos, cmd, cmd_args)
+
+    x_coord_idxs, y_coord_idxs = svg_meta.cmd_coords(cmd)
+    cmd_args = list(cmd_args)  # we'd like to mutate
+    new_x, new_y = new_endpoint
+    if cmd.islower():
+        new_x = new_x - curr_pos.x
```

Comment:
I believe this subtraction operation may introduce futher float numerical imprecision..
That's the reason why in my patch I opted to only do this endpoint adjustment inside the absolute() method -- which, by the way, is the origin of the float imprecision, because of the addition of floats to make relative coords absolute.

Reply:
Fair point. I'm tempted to make snap to end its own thing because only doing it in absolute feels odd to me. I think for now I won't :)